### PR TITLE
clamp Categorical logit from -inf to min_fifo when calculating entropy

### LINKED ISSUE #40553: Categorical entropy of logits is inconsistent with probs
## 🐛 Bug

When using the `Categorical` distribution, having 0-probability classes results in correct entropy calculations, while using the same distribution as logits (which includes `-inf`) does not.

## To Reproduce

```
In [1]: import torch; from torch.distributions import Categorical

In [2]: logits = torch.tensor([0, 0, -float('inf')])

In [3]: Categorical(logits=logits).entropy()
Out[3]: tensor(nan)

In [4]: Categorical(probs=logits.softmax(-1)).entropy()
Out[4]: tensor(0.6931)
```

## Expected behavior

Entropy should be the same in both cases.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5.0
 - OS (e.g., Linux):  Mac OSX 10.14.6
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information:

## Additional context

I hope this is a simple fix?

cc @ezyang @gchanan @zou3519 @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

### PR DESCRIPTION
**Pull request by mattip**
Fixes gh-40553 by clamping logit values when calculating Categorical.entropy

### CODE REVIEW

**Comment by ngimel** [comment]
Flake errors are real. 
@fritzo, @neerajprad do the changes look good to you otherwise?

File: torch/distributions/categorical.py [hunk 1]
```diff
@@ -115,7 +116,9 @@ def log_prob(self, value):
         return log_pmf.gather(-1, value).squeeze(-1)
 
     def entropy(self):
-        p_log_p = self.logits * self.probs
+        min_real = torch.finfo(self.logits.dtype).min
+        logits = torch.clamp(self.logits, min=min_real)
+        p_log_p = logits * self.probs
```
Comment (neerajprad) [inline torch/distributions/categorical.py:121]: This will be a good place to use [xlogy](https://github.com/pytorch/pytorch/issues/22656) when available.

File: torch/distributions/categorical.py [hunk 2]
```diff
@@ -51,6 +51,7 @@ def __init__(self, probs=None, logits=None, validate_args=None):
         else:
             if logits.dim() < 1:
                 raise ValueError("`logits` parameter must be at least one-dimensional.")
+            # Convert -inf to a real number
```
[the remark this answers is not in the document]
Comment (mattip) [reply torch/distributions/categorical.py:54]: I meant "real" as "non-inf".
```suggestion
            # Normalize -inf to min_real
```

File: torch/distributions/categorical.py (hunk 1 above)
Comment (mattip) [reply torch/distributions/categorical.py:121]: I think xlogy calculates `x * log(y)` where here we want `x * y`

Comment (neerajprad) [reply torch/distributions/categorical.py:121]: That's right, we are using the precomputed probs / logits attributes directly.

File: torch/distributions/categorical.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (neerajprad) [reply torch/distributions/categorical.py:54]: I meant that normalization below would  retain `-inf` (indices with 0 prob).

```python
>>> t = torch.tensor([float('-inf'), 2])
>>> t - t.logsumexp(dim=-1, keepdim=True)
tensor([-inf, 0.])
```

[the remark this answers is not in the document]
Comment (fritzo) [reply torch/distributions/categorical.py:54]: So the comment should be merely "Normalize", right?

**Review by fritzo — COMMENTED** [review]
LGTM except for the comment on line 54

File: torch/distributions/categorical.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (neerajprad) [reply torch/distributions/categorical.py:54]: That's right, I just wanted to also make sure that we aren't assuming that the logits will get clipped here, but as far as I can see, entropy is the only place where `-inf` values could cause issues and that has been fixed separately.

[the remark this answers is not in the document]
Comment (mattip) [reply torch/distributions/categorical.py:54]: What would you prefer the comment read? "Normalize -inf in logit"?

[the remark this answers is not in the document]
Comment (fritzo) [reply torch/distributions/categorical.py:54]: Simply "Normalize"

Comment (mattip) [inline torch/distributions/categorical.py:54]:
```suggestion
            # Normalize
```

**Review by fritzo — APPROVED** [review]

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
I see a related error on internal tests:
```
stderr:
test_entropy (test_distributions.TestJit) ... caffe2/torch/csrc/jit/ir/node_hashing.cpp:203:42: runtime error: -1.79769e+308 is outside the range of representable values of type 'float'
    #0 0x7ff8fec85d97 in torch::jit::HashNode::operator()(torch::jit::Node const*) const (/mnt/xarfuse/uid-30041/b83ac162-ns-4026533886/libomnibus.so+0x7013bd97)
    #1 0x7ff8fe5cd2f2 in std::pair<std::__detail::_Node_iterator<torch::jit::Node*, true, true>, bool> std::_Hashtable<torch::jit::Node*, torch::jit::Node*, std::allocator<torch::jit::Node*>, std::__detail::_Identity, torch::jit::EqualNode, torch::jit::HashNode, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, true, true> >::_M_insert<torch::jit::Node* const&, std::__detail::_AllocNode<std::allocator<std::__detail::_Hash_node<torch::jit::Node*, true> > > >(torch::jit::Node* const&, std::__detail::_AllocNode<std::allocator<std::__detail::_Hash_node<torch::jit::Node*, true> > > const&, std::integral_constant<bool, true>) (/mnt/xarfuse/uid-30041/b83ac162-ns-4026533886/libomnibus.so+0x6fa832f2)
    #2 0x7ff8fe5e3c1c in torch::jit::(anonymous namespace)::ConstantPooling(torch::jit::Block*, std::unordered_set<torch::jit::Node*, torch::jit::HashNode, torch::jit::EqualNode, std::allocator<torch::jit::Node*> >&, torch::jit::AliasDb const&) (/mnt/xarfuse/uid-30041/b83ac162-ns-4026533886/libomnibus.so+0x6fa99c1c)
    #3 0x7ff8fe5e33d6 in torch::jit::ConstantPooling(std::shared_ptr<torch::jit::Graph> const&) (/mnt/xarfuse/uid-30041/b83ac162-ns-4026533886/libomnibus.so+0x6fa993d6)
    #4 0x7ff8fe79c93a in torch::jit::GraphFunction::optimized_graph() const (/mnt/xarfuse/uid-30041/b83ac162-ns-4026533886/libomnibus.so+0x6fc5293a)
```
This is a jit version of test_entropy which was likely touched by this diff, I don't know why OSS CI did not catch it.

**Comment by fritzo** [comment]
Do you understand why that error is being triggered? It looks as if `torch.finfo()` sees `self.logits.dtype` as `torch.double`, but that `self.logits.clamp(...)` is using type `torch.float` 😕

**Comment by ngimel** [comment]
I don't, maybe jit script does not correctly script finfo, or TestJit is using default double datatype? cc @suo, @eellison for internal jit failure.

**Comment by ngimel** [comment]
@mattip can you please add the following to your diff
```
diff --git a/pytorch/aten/src/ATen/core/ivalue.cpp b/fbcode/caffe2/aten/src/ATen/core/ivalue.cpp
--- a/pytorch/aten/src/ATen/core/ivalue.cpp
+++ b/pytorch/aten/src/ATen/core/ivalue.cpp
@@ -360,7 +360,7 @@
     case IValue::Tag::Double: {
       double d = v.toDouble();
       int c = std::fpclassify(d);
-      if (c == FP_NORMAL || c == FP_ZERO) {
+      if ((c == FP_NORMAL || c == FP_ZERO) && std::abs(d) < 1e10) {
         int64_t i = int64_t(d);
         if (double(i) == d) {
           return out << i << ".";
diff --git a/pytorch/torch/csrc/jit/ir/node_hashing.cpp b/fbcode/caffe2/torch/csrc/jit/ir/node_hashing.cpp
--- a/pytorch/torch/csrc/jit/ir/node_hashing.cpp
+++ b/pytorch/torch/csrc/jit/ir/node_hashing.cpp
@@ -200,7 +200,7 @@
     } else if (
         type->isSubtypeOf(NumberType::get()) &&
         k->kindOf(attr::value) == AttributeKind::f) {
-      constant_hash = std::hash<float>{}(k->f(attr::value));
+      constant_hash = std::hash<double>{}(k->f(attr::value));
     } else if (type->isSubtypeOf(BoolType::get())) {
       constant_hash = std::hash<bool>{}(k->i(attr::value));
     }
```
This solves internal test failures, but I want to make sure that OSS CI is ok with this too. Thanks!

**Comment by mattip** [comment]
> This solves internal test failures, but I want to make sure that OSS CI is ok

@ngimel From my limited understanding the changes seem OK. I applied the patch. It would be good if we could also at some point add a test for this code path, maybe there are more like this lurking around

**Comment by mattip** [comment]
It seems there are merge conflicts. Rebased to clear them.

**Comment by ngimel** [comment]
These failures were coming from running regular tests under ASAN, we have an asan build in OSS CI, but for some reason it's not triggering those.
